### ISSUE TITLE
[Feature request] Arbitrary base learner

### ISSUE THREAD

**Issue by zachmayer** [dmlc/xgboost#5802]
Its pretty cool that I can define my own loss function and gradient for xgboost, and then use the linear, tree, or dart base learners to optimize my loss function.

It'd be really cool if I could specify my own base learner, perhaps in the form of an sklearn class with a fit method, a predict method, and support for sample weights.

It'd really open up a whole new world of possibilities to be able to use the Xgboost algorithm to fit a wider range of possible base learners.

**Comment by hcho3** [comment]
@zachmayer Is `StackingClassifier` and `StackingRegressor` an option for you? We recently added support for it: #5780

**Comment by hcho3** [comment]
Oops, my bad. When you say  "base learner," you mean that you want a custom model as part of the ensemble?

**Comment by zachmayer** [comment]
Yes exactly.

So for example if I wanted to boost a kernel Svm I could do that

**Comment by tunguz** [comment]
Here is a related issue that has just been opened: 

https://github.com/rapidsai/cuml/issues/2435

Adding AfaBoost to cuml might be a good stop-gap measure.

**Comment by zachmayer** [comment]
sklearn adaboost already supports arbitrary base learners: `sklearn.ensemble.AdaBoostClassifier`

XGboost is way better than adaboost though, and supports a bunch of features adaboost doesn't have:
1. You can specify an arbitrary loss function in xgboost.
2. You can specify a gradient for your loss function, and use the gradient in your base learner.
3. You can specify an arbitrary evaluation function in xgboost.
4. You can do early stopping with xgboost.
5. You can run xgboost base learners in parallel, to mix "random forest" type learning with "boosting" type learning

**Comment by zachmayer** [comment]
@tunguz "Run adaboost on a gpu" isn't really what I'm looking for.

"Run adaboost with an arbitrary base learner, arbitrary loss function, arbitrary gradient, arbitrary evaluation, early stopping, and a mix of parallel learners (aka bagging) and boosting" would suit my needs, but that's another way to say "run xgboost with an arbitrary base learner" 😁

**Comment by zachmayer** [comment]
Just a follow up on this:
- [ngboost](https://github.com/stanfordmlgroup/ngboost) supports arbitrary base learners, which solves the problem for me for now.
- There's an interesting new package called [Grownet](https://github.com/sbadirli/GrowNet) which has some evidence that boosting different weak learners (specifically neural networks) is useful. (There's a [paper](https://arxiv.org/pdf/2002.07971.pdf) too)
